boba_community/turing-monsters/contracts/NFTMonsterV2.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/interfaces/IERC2981.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Burnable.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Pausable.sol";
+import "./RandomlyAssigned.sol";
+import "./WithRecover.sol";
+import "./WithOnChainMetaData.sol";
+
+
+contract NFTMonsterV2 is IERC2981, ERC721Burnable, ERC721Pausable, RandomlyAssigned, WithRecover, WithOnChainMetaData {
+
+    uint256 public constant PRICE = 0.0000000001 ether;
+    uint256 public constant ROYALTY_PERCENTAGE = 5; // 5 %
+    uint256 public constant MAX_MINT_IN_PUBLIC = 3;
+    address[] public projectOwners;
+
+    mapping(address => uint256) public amountMintedInPublicSale;
+
+    event MintedNFT(uint256 indexed id);
+
+    constructor(string memory name_, string memory symbol_, uint256 maxNFTs_,
+        address[] memory creatorAddresses_, address turingHelperAddress_)
+    ERC721(name_, symbol_)
+    WithOnChainMetaData(turingHelperAddress_)
+    RandomlyAssigned(maxNFTs_, 0, turingHelperAddress_) // Max. x NFTs available; Start counting from 0
+    {
+        _pause();
+        projectOwners = creatorAddresses_;
+    }
+
+    modifier saleIsOpen {
+        require(tokenCount() <= totalSupply(), "Sale end");
+        if (_msgSender() != owner()) {
+            require(!paused(), "Pausable: paused");
+        }
+        _;
+    }
+
+    /// @notice Called with the sale price to determine how much royalty
+    //          is owed and to whom.
+    /// @param _tokenId - the NFT asset queried for royalty information
+    /// @param _salePrice - the sale price of the NFT asset specified by _tokenId
+    /// @return receiver - address of who should be sent the royalty payment
+    /// @return royaltyAmount - the royalty payment amount for _salePrice
+    function royaltyInfo(
+        uint256 _tokenId,
+        uint256 _salePrice
+    ) external override view returns (
+        address receiver,
+        uint256 royaltyAmount
+    ) {
+        return (address(this), (_salePrice * ROYALTY_PERCENTAGE)/100);
+    }
+
+    /**
+    * For DApp UX to determine how many NFTs can be minted.
+    */
+    function allowedMintAmount(address account_) view external returns (uint256) {
```

Comment:
minor: not much gas increase, but eager to know your thoughts around calculation on the UI side, since both are public?

Reply:
What do you mean with that? :-)